python/lsst/scarlet/lite/operators_pybind11.cc

```diff
@@ -82,65 +83,65 @@ void apply_filter(
 // and create a boolean map of "orphans" that are non-monotonic in all directions.
 template <typename M>
 void get_valid_monotonic_pixels(
-    const int i,
-    const int j,
+    const int start_i,
+    const int start_j,
     Eigen::Ref<M, 0, Eigen::Stride<Eigen::Dynamic, Eigen::Dynamic>> image,
     Eigen::Ref<MatrixB, 0, Eigen::Stride<Eigen::Dynamic, Eigen::Dynamic>> unchecked,
     Eigen::Ref<MatrixB, 0, Eigen::Stride<Eigen::Dynamic, Eigen::Dynamic>> orphans,
     const double variance,
     Eigen::Ref<Bounds, 0, Eigen::Stride<4, 1>> bounds,
     const double thresh=0
-){
-    // Check the pixel below this one
-    if(i>0 && unchecked(i-1,j)){
-        if(image(i-1,j) <= image(i,j)+variance && image(i-1,j) > thresh){
-            unchecked(i-1,j) = false;
-            orphans(i-1,j) = false;
-            if(i-1 < bounds(0)){
-                bounds(0) = i-1;
-            }
-            get_valid_monotonic_pixels(i-1, j, image, unchecked, orphans, variance, bounds);
-        } else {
-            orphans(i-1,j) = true;
-        }
-    }
-    // Check the pixel above this one
-    if(i < image.rows()-1 && unchecked(i+1,j)){
-        if(image(i+1,j) <= image(i,j)+variance && image(i+1,j) > thresh){
-            unchecked(i+1,j) = false;
-            orphans(i+1,j) = false;
-            if(i+1 > bounds(1)){
-                bounds(1) = i+1;
+) {
+    // Create a queue to store pixels that need to be processed
+    std::queue<std::pair<int, int>> pixel_queue;
+    pixel_queue.push({start_i, start_j});
+
+    while (!pixel_queue.empty()) {
+        auto [i, j] = pixel_queue.front();
+        pixel_queue.pop();
+
+        // Define directions: down, up, left, right
+        const int di[] = {-1, 1, 0, 0};
```

Comment:
While this might get optimized away, these shouldn't be in the while loop or even in a templated function - they can go outside the function.